pyrsa/data/noise.py

```diff
@@ -0,0 +1,78 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+"""
+Created on Mon Apr 13 12:42:06 2020
+
+@author: heiko
+"""
+
+from collections.abc import Iterable
+import numpy as np
+
+
+def cov_from_residuals(residuals, dof=None):
+    """
+    computes an optimal shrinkage estimate of the precision matrix from
+    the residuals as described by Ledoit and Wolfe (2004): "A well-conditioned
+    estimator for large-dimensional covariance matrices"
+
+    Args:
+        residuals(numpy.ndarray or list of these): n_residuals x n_channels
+            matrix of residuals
+        dof(int or list of int): degrees of freedom for covariance estimation
+            defaults to n_res - 1, should be corrected for the number
+            of regressors in a GLM if applicable.
+
+    Returns:
+        numpy.ndarray (or list): sigma_p: covariance matrix over channels
+
+    """
+    if not isinstance(residuals, np.ndarray) or len(residuals.shape) > 2:
+        s_shrink = []
+        for i in range(len(residuals)):
+            if dof is None:
+                s_shrink.append(cov_from_residuals(residuals[i]))
+            elif isinstance(dof, Iterable):
+                s_shrink.append(cov_from_residuals(residuals[i], dof[i]))
+            else:
+                s_shrink.append(cov_from_residuals(residuals[i], dof))
+    else:
+        if dof is None:
+            dof = residuals.shape[0] - 1
+        residuals = residuals - np.mean(residuals, axis=0, keepdims=True)
```

Comment:
you could describe the steps of the algebraic shrinkage transform to keep consistency (also a sanity check that you followed Ledoit and Wolf's method correctly)

Reply:
done